### PR TITLE
fix: Detect and handle circular `$ref` in `_dereference_schema`

### LINKED ISSUE #3870: `RecursionError` in `_to_gemini_schema()` when processing circular `$ref` in `$defs`
**Describe the bug**
The `_to_gemini_schema()` function in `google.adk.tools._gemini_schema_util` triggers a `RecursionError` when processing JSON Schemas with circular `$ref` references in `$defs`. The `_dereference_schema()` helper function recursively resolves these references without detecting circular dependencies, causing infinite recursion.

**To Reproduce**
- Minimal code to reproduce the issue:
    ```
    from google.adk.tools._gemini_schema_util import _to_gemini_schema
    
    # Schema with circular reference: Attendee -> alternateAttendee -> Attendee
    schema_with_circular_refs = {
        "type": "object",
        "properties": {
            "body": {
                "type": "object",
                "properties": {
                    "attendees": {
                        "type": "array",
                        "items": {"$ref": "#/$defs/Attendee"}
                    },
                    "end": {"$ref": "#/$defs/DateTimeTimeZone"}
                }
            }
        },
        "$defs": {
            "DateTimeTimeZone": {
                "type": "object",
                "properties": {
                    "dateTime": {"type": "string"},
                    "timeZone": {"type": "string"}
                }
            },
            "Attendee": {
                "type": "object",
                "properties": {
                    "proposedNewTime": {
                        "type": "object",
                        "properties": {
                            "start": {"$ref": "#/$defs/DateTimeTimeZone"},
                            "end": {"$ref": "#/$defs/DateTimeTimeZone"}
                        }
                    },
                    "alternateAttendee": {"$ref": "#/$defs/Attendee"}
                }
            }
        },
        "required": ["body"]
    }
    
    # This will trigger a RecursionError
    gemini_schema = _to_gemini_schema(schema_with_circular_refs)
    ```
- Steps to reproduce:
   1. Install `google-adk` and `google-genai`
   2. Run the code above
   3. Observe `RecursionError: maximum recursion depth exceeded`

**Expected behavior**
- The function should either:
   1. Detect circular references and raise a clear, descriptive error (e.g., ValueError: Circular reference detected in schema at path ...)
   2. Handle circular references gracefully by tracking visited definitions during dereferencing
   
- Stack trace:
   ```
      [...]
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 94, in _resolve_refs
        return _resolve_refs(resolved)
               ^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 94, in _resolve_refs
        return _resolve_refs(resolved)
               ^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 94, in _resolve_refs
        return _resolve_refs(resolved)
               ^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      [Previous line repeated 1 more time]
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 94, in _resolve_refs
        return _resolve_refs(resolved)
               ^^^^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
      File "/home/ubuntu/venv/lib/python3.12/site-packages/google/adk/tools/_gemini_schema_util.py", line 100, in _resolve_refs
        return {key: _resolve_refs(value) for key, value in sub_schema.items()}
                     ^^^^^^^^^^^^^^^^^^^^
    RecursionError: maximum recursion depth exceeded
   ```

**Desktop (please complete the following information):**
 - OS: Linux (Ubuntu)
 - Python version(python -V): Python 3.12.3
 - ADK version(pip show google-adk):
    - `google-adk==1.20.0`
    - `google-genai==1.54.0`

 **Model Information:**
 - Are you using LiteLLM: No
 - Which model is being used(e.g. gemini-2.5-pro): N/A (error occurs during schema conversion, before model invocation)

**Additional Context**
 This issue was discovered while building an MCP server tool that uses schemas with legitimate circular references for recursive data structures (for example, attendees with alternate attendees)
- The root cause appears to be in `_dereference_schema()`, which currently does not perform cycle detection when recursively resolving `$ref` pointers in `$defs`
- A previous attempt was made to fix this in `google-genai`, but that change could not be merged because the affected file (`types.py`) is dynamically generated
- I was advised to raise an issue (and potentially a PR) here; once I figure out an appropriate approach, I am happy to follow up with a PR

### PR DESCRIPTION
**Pull request by foivos-all**
### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #3870
- Related: #3870

**Problem:**
The `_dereference_schema()` function in `google.adk.tools._gemini_schema_util` triggers a `RecursionError` when processing JSON schemas with circular `$ref` references. This occurs because the function recursively resolves references without detecting circular dependencies, causing infinite recursion when a schema references itself (e.g., a tree structure where a node can contain child nodes of the same type).

Additionally, the original implementation only handled `$defs`-style references (`#/$defs/TypeName`) but not JSON Pointer references (`#/properties/path/to/field`), which are commonly used in OpenAPI and other schema formats.

**Solution:**
Enhanced `_dereference_schema()` with the following improvements:

1. **Circular reference detection**: Track references currently being resolved using a `resolving` set. When a circular reference is detected, return a placeholder schema (`{"type": "object", "description": "Circular reference to <ref>"}`) instead of recursing infinitely.

2. **JSON Pointer support**: Added `_resolve_json_pointer()` helper function to resolve references that use JSON Pointer syntax (`#/properties/...`), in addition to existing `$defs` support.

This allows for a graceful degradation, where rather than raising an error, circular references are replaced with valid placeholder schemas, allowing tools with recursive data structures to function correctly.

### Testing Plan

**Unit Tests:**

- [x] I have added or updated unit tests for my change
   - Added new test file with 18 test cases covering different circular reference patterns
- [x] All unit tests pass locally.
   - 4026 passed
   - 2169 warnings
   - Took 107.06s

**Manual End-to-End (E2E) Tests:**
1. Schema with circular `$defs` reference
    ```
    from google.adk.tools._gemini_schema_util import _to_gemini_schema
    
    schema = {
        "type": "object",
        "properties": {
            "attendees": {
                "type": "array",
                "items": {"$ref": "#/$defs/Attendee"}
            }
        },
        "$defs": {
            "Attendee": {
                "type": "object",
                "properties": {
                    "name": {"type": "string"},
                    "alternateAttendee": {"$ref": "#/$defs/Attendee"} # Circular
                }
            }
        }
    }
    
    gemini_schema = _to_gemini_schema(schema)
    ```
    - **Before fix:** `RecursionError`
    - **After fix:** Success

2. Schema with JSON Pointer circular reference
    ```
    from google.adk.tools._gemini_schema_util import _to_gemini_schema
    
    schema = {
        "type": "object",
        "properties": {
            "body": {
                "type": "object",
                "properties": {
                    "attendees": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "start": {
                                    "$ref": "#/properties/body/properties/end" # JSON Pointer
                                }
                            }
                        }
                    },
                    "end": {"type": "object", "properties": {"dateTime": {"type": "string"}}}
                }
            }
        }
    }
    
    gemini_schema = _to_gemini_schema(schema)
    ```
    - **Before fix:** `RecursionError`
    - **After fix:** Success

3. Schema with 2 circular `$defs` reference
    ```
    from google.adk.tools._gemini_schema_util import _to_gemini_schema
    
    schema = {
        "type": "object",
        "properties": {
            "attendees": {
                "type": "array",
                "items": {"$ref": "#/$defs/Attendee"}
            },
            "organizer": {"$ref": "#/$defs/Person"}
        },
        "$defs": {
            "Attendee": {
                "type": "object",
                "properties": {
                    "name": {"type": "string"},
                    "alternateAttendee": {"$ref": "#/$defs/Attendee"}  # Circular 1
                }
            },
            "Person": {
                "type": "object",
                "properties": {
                    "name": {"type": "string"},
                    "manager": {"$ref": "#/$defs/Person"}  # Circular 2
                }
            }
        }
    }
    
    gemini_schema = _to_gemini_schema(schema)
    ```
    - **Before fix:** `RecursionError`
    - **After fix:** Success

### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

### Additional context

**Note on google-genai dependency:**
While this fix resolves the issue for ADK's schema conversion pipeline, the `google-genai` library's `Schema.from_json_schema()` method has a similar issue in its `convert_json_schema()` function. However, since ADK calls `_dereference_schema()` before passing schemas to `google-genai`, this fix protects ADK users from the `RecursionError`. Direct users of `google-genai` would need a similar fix in that library's auto-generated `types.py` file.

**Applicability:**
This fix enables building MCP server tools that wrap complex APIs  which often contain recursive schema definitions for representing hierarchical data structures.

### CODE REVIEW

**Comment by rkrumins** [comment]
Great job on this @foivos-all and thank you very much! I'm being affected by the same issue using various MCPs together with ADK. Very keen on getting this merged given that resolves the circular $ref issue. 🙏

**Comment by rohityan** [comment]
Hi @foivos-all ,Thank you for your contribution! We appreciate you taking the time to submit this pull request. Your PR has been received by the team and is currently under review. We will provide feedback as soon as we have an update to share.

**Comment by rohityan** [comment]
Hi @seanzhou1023 , can you please review this.

**Comment by subbuj9** [comment]
@foivos-all I merged the changes and verified but the fix is not working in my scenario. Below is the some part of MCP server response payload.
{
	"name": "createSubscriber",
	"description": "If a subscribe request is received .....",
	"inputSchema": {
		"type": "object",
		"properties": {
			"subsRequest": {
				"**$defs**": {
					"KnJsonAliasIdInfo": {
						"type": "object",
						"properties": {
							"aliasId": {
								"type": "string"
							},
							"aliasIdIssuer": {
								"type": "string"
							},
							"aliasIdType": {
								"type": "string"
							}
						}
					},
				},
				"type": "object",
				"properties": {
					"accountId": {
						"type": "string",
						"description": "Value to uniquely identify the account."
					},
					"addPackageInfo": {
						"**$ref**": "#/$defs/KnJsonPackageInfo",
						"description": "..."
					},
					"affiliateId": {
						"type": "string",
						"description": "Indicates the entity sending the request."
					},
					"aliasIdList": {
						"description": "...",
						"type": "array",
						"items": {
							"$ref": "#/$defs/KnJsonAliasIdInfo"
						}
					},
				},
				"required": [
					"mdn"
				]
			},
			"headers": {
				"type": "object"
			}
		},
		"required": [
			"subsRequest",
			"headers"
		],
		"additionalProperties": false
	}
} 


Observing the below error 

2025-12-17 13:32:25,230 - ERROR - adk_web_server.py:1509 - Error in event_generator: 'NoneType' object is not subscriptable
Traceback (most recent call last):
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\cli\adk_web_server.py", line 1499, in event_generator
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\runners.py", line 505, in run_async
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\runners.py", line 493, in _run_with_trace
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\runners.py", line 722, in _exec_with_plugin
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\runners.py", line 482, in execute
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\agents\base_agent.py", line 294, in run_async
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\agents\llm_agent.py", line 460, in _run_async_impl
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\flows\llm_flows\base_llm_flow.py", line 370, in run_async
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\flows\llm_flows\base_llm_flow.py", line 389, in _run_one_step_async
    async for event in agen:
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\flows\llm_flows\base_llm_flow.py", line 507, in _preprocess_async
    await tool.process_llm_request(
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\tools\base_tool.py", line 129, in process_llm_request
    llm_request.append_tools([self])
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\models\llm_request.py", line 255, in append_tools
    declaration = tool._get_declaration()
                  ^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\tools\mcp_tool\mcp_tool.py", line 119, in _get_declaration
    parameters = _to_gemini_schema(input_schema)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\adk\tools\_gemini_schema_util.py", line 178, in _to_gemini_schema
    return Schema.from_json_schema(
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\genai\types.py", line 2692, in from_json_schema
    return convert_json_schema(
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\genai\types.py", line 2639, in convert_json_schema
    key: convert_json_schema(
         ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\genai\types.py", line 2639, in convert_json_schema
    key: convert_json_schema(
         ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\XFJP78\AppData\Local\anaconda3\Lib\site-packages\google\genai\types.py", line 2553, in convert_json_schema
    json_schema_dict = _resolve_ref(
    
    **_TypeError: 'NoneType' object is not subscriptable_**

**Comment by foivos-all** [comment]
@subbuj9 Thanks for reporting. While nested `$defs` are technically allowed per the spec, your schema seems to have some issues:
- Your `$defs` is located at `#/properties/subsRequest/$defs`, but you're using `#/$defs/` which points to the root level. References should be `#/properties/subsRequest/$defs/KnJsonAliasIdInfo`, since that is where you're defining it
- You're also referencing `#/$defs/KnJsonPackageInfo` but I can't see this definition in your schema
- Again, while nested `$defs` are technically allowed the JSON Schema maintainers warn you might lose meta-schema validation at nested locations, making debugging harder

I'm assuming the `NoneType` error occurs because my code looks for `$defs` at the root level where your references point, but isn't finding anything. I would suggest fixing the schema and trying again. Can you try:
- Moving `$defs` to the root level, or
- Updating all `$ref` paths to match the actual nested location

**Comment by rohityan** [comment]
Hi @foivos-all , we appreciate your patience and support. Can you please fix the failing unit tests.

**Comment by foivos-all** [comment]
Hi @ryanaiagent, I merged the `main` branch into my branch and ran the tests locally. All tests passed:
image
It says:
- 3938 passed
- 2102 warnings

One thing I observed is that the number of unit tests from my local run to the CI one is different. Hope this helps!

**Comment by rohityan** [comment]
Hi @wukath , can you please review this.

**Comment by wukath** [comment]
thanks for contributing! could you add unit tests as well?

**Comment by foivos-all** [comment]
Hi @wukath, I've added the unit tests with a new test file has 18 test cases covering different circular reference patterns (linked lists, trees, deep chains, composition operators like `anyOf`/`allOf`/`oneOf`, and edge cases). I also included a test that directly reproduces the scenario from the issue I was referencing (#3870)

I ran the new tests before and after my changes:
- Before: 13 failed with `RecursionError`, 5 passed
- After: 18 passed, 0 failed

Hope this confirms the circular reference issues are fixed and nothing broke for non-circular cases. I've also updated the PR description to better reflect these changes.

File: src/google/adk/tools/_gemini_schema_util.py [hunk 1]
```diff
@@ -107,11 +107,47 @@ def _dereference_schema(schema: dict[str, Any]) -> dict[str, Any]:
...
+    parts = ref_path[2:].split("/")
+    current = root
+
+    # Traverse the schema following the path.
+    for part in parts:
+      if not isinstance(current, dict):
+        return None
+      current = current.get(part)
+      if current is None:
+        return None
```
[the remark this answers is not in the document]
Comment (foivos-all) [reply src/google/adk/tools/_gemini_schema_util.py:135]: Added to make the JSON pointer handling more robust

File: tests/unittests/tools/test_circular_refs.py [hunk 2]
```diff
@@ -0,0 +1,383 @@
+from google.adk.tools._gemini_schema_util import _dereference_schema
+
+
+class TestDereferenceSchemaCircularRefs:
+  """Test circular $ref detection and handling in _dereference_schema."""
```
[the remark this answers is not in the document]
Comment (foivos-all) [reply tests/unittests/tools/test_circular_refs.py:19]: Added the extra tests as well

**Comment by rohityan** [comment]
Hi @foivos-all , could you address the recommendations.

**Comment by foivos-all** [comment]
Hey @ryanaiagent, I already addressed them. I enhanced the JSON pointer handling, as Gemini suggested, to make it more robust and also added a few more tests

**Comment by rohityan** [comment]
Hi @foivos-all, Your PR has been received by the team and is currently under review. We will provide feedback as soon as we have an update to share.

**Comment by rohityan** [comment]
Hi @GWeale , can you please review this.

**Comment by foivos-all** [comment]
Made the necessary changes to fix the `mypy` error caused by a generic `dict` type annotation

**Comment by foivos-all** [comment]
@ryanaiagent Any updates on this?

**Comment by foivos-all** [comment]
@rohityan I noticed that a similar version to my code was merged from an internal [source](https://github.com/google/adk-python/commit/7de5bc54e11986f70a48a9dd83ea39be58ebce40#diff-8a4df9a99e38d640538533c23a51d0f5455352664ff2424706d4cb57d196b1ad), which is great! However, I noticed the JSON Pointer resolution feature is still missing from the current implementation. I originally added this to handle `$ref` references like `#/properties/address` which go beyond just `$defs` resolution. I believe this would be a valuable addition to fully support OpenAPI 3.1 schemas that use JSON Pointer references. Would it be possible to add this feature as well?

**Comment by GWeale** [comment]
Thanks for the thorough work here! The core RecursionError fix for circular $ref landed on main via an equivalent internal change (commit 7de5bc54 in tools/_gemini_schema_util.py), which closed #3870, so closing this PR. As you noted in the thread, the JSON Pointer (#/properties/...) resolution piece isn't in main yet — if you're still interested, a small focused PR just for that would be very welcome.

**Comment by foivos-all** [comment]
Thank you for responding @GWeale. As per your request, I've opened a new PR, specifically for the JSON pointer fix. I would really appreciate it if you could review it: https://github.com/google/adk-python/pull/6148
